packages/react-compose/src/next/README.md

```diff
@@ -0,0 +1,226 @@
+# @fluentui/react-compose
+
+`@fluentui/react-compose` is a library of utilities for composing Fluent UI components.
+
+## A basic component walkthrough
+
+Building a recomposable component requires that we build legos; we put them together, but we let others put them together how they may need. Here's what's needed:
+
+- **State hooks** - all processing of the component should be done within one or more reusable hooks. The hooks should be allowed to manipulate the state in some reusable way, so that recomposing the component can be easily done.
+- **Render function** - a function which takes in final state and returns JSX. (e.g. `renderButton`)
+- **Component factory function** - a create function which returns `{ state, render }` parts for creating the component. This is used to scaffold a new version of the component as needed.
+- **Style hooks** - hooks which can take in styles and provide appropriate classnames to the anatomy of the component.
+
+With these building blocks, you can compose or recompose the component in numerous ways.
+
+### Simple example
+
+A render function:
+
+```jsx
+const renderButton = state => {
+  return <button {...state} />;
+};
+```
+
+A hook which can manipulate the state (defining accessibility and behaviors):
+
+```jsx
+const useButton = state => {
+  // If the button is rendered as something besides a button or anchor, make it focusable.
+  if (state.as !== 'button' && state.as !== 'a') {
+    state.tabIndex = 0;
+  }
+};
+```
+
+A factory function sets up the state and render function:
+
+```jsx
+const createButton = (userProps, ref, defaultProps) => {
+  const state = _.merge({}, defaultProps, userProps);
+
+  // Apply button behaviors.
+  useButton(state);
+
+  return { state, render };
+};
+```
+
+A button can now be easily scaffolded, along with your choice of styling systems:
+
+```jsx
+import * as styles from './Button.scss';
+
+const useStyles = makeStyles(styles);
+
+const Button = React.forwardRef((props, ref) => {
+  const { state, render } = createButton(props, ref);
+
+  // Inject classNames as needed.
+  useStyles(state);
+
+  return render(state);
+});
+```
+
+### Details
+
+#### Creating mutable state with `mergeProps`
+
+In the previous example, `_.merge` was used to deep clone the props into a state object. Creating a single clone and using that to construct state simplifies hook development and usage; rather than trying to re-clone objects unnecessarily on every small mutation, hooks can assume operating against a draft state. This creates more self contained hooks, which can ensure they are used correctly, avoiding accidents like stomping on existing event handlers by blind object assigning the results.
```

Comment:
> Creating a single clone and using that to construct state simplifies hook development and usage;

Why? Do you have examples in other libraries?
For me state mutations are confusing and non-intuitive.

> rather than trying to re-clone objects unnecessarily on every small mutation

Is it a performance bottleneck?

Reply:
We have components like `Tree`, `Table`, etc. that accept nested shorthands:

```tsx
[
  {
    id: "tree-item-1",
    title: "House Lannister",
    items: [
      {
        id: "tree-item-11",
        title: "Tywin",
        items: [
          {
            id: "tree-item-111",
            title: "Jaime"
          },
          {
            id: "tree-item-112",
            title: "Cersei"
          },
          {
            id: "tree-item-113",
            title: "Tyrion"
          }
        ]
      },
      {
        id: "tree-item-12",
        title: "Kevan",
        items: [
          {
            id: "tree-item-121",
            title: "Lancel"
          },
          {
            id: "tree-item-122",
            title: "Willem"
          },
          {
            id: "tree-item-123",
            title: "Martyn"
          }
        ]
      }
    ]
  },
  {
    id: "tree-item-2",
    title: "House Targaryen",
    items: [
      {
        id: "tree-item-21",
        title: "Aerys",
        items: [
          {
            id: "tree-item-211",
            title: "Rhaegar"
          },
          {
            id: "tree-item-212",
            title: "Viserys"
          },
          {
            id: "tree-item-213",
            title: "Daenerys"
          }
        ]
      }
    ]
  }
];
```

https://fluentsite.z22.web.core.windows.net/components/tree/definition?showCode=true&showRtl=false&showTransparent=false&showVariables=false#types-tree

Will be deep clone more effective in this case?

Reply:
This for inspiration: https://immerjs.github.io/immer/docs/introduction

arrays will not be traversed, so your example would have created an object clone and the items would have been left unchanged.

I can see this being confusing. I think blind deep clones aren't good enough, and shallow clones are error prone as well. That's why I felt the need to have something besides a dumb deep merge that would do something a little more specific for draft state.

Reply:
And yes; immutability does add overhead. This is what led me here. Trying to kill extra _rest, _merge, etc calls showing up in flamegraphs.

Reply:
And also, one size doesn't fit all. If the deep clone does or does not do something you need done differently, we can either tweak the utility or use an alternative approach.

Created an RFC to discuss mutable approach vs immutable approach:
https://github.com/microsoft/fluentui/issues/14365

Also have scheduled time on friday to discuss :)